src/uct/tcp/tcpcm/tcpcm_ep.c

```diff
@@ -0,0 +1,298 @@
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    if (self->is_on_pending) {
+        ucs_list_del(&self->list_elem);
+        self->is_on_pending = 0;
+    }
+
+    /* remove the slow progress function in case it was placed on the slow progress
+     * chain but wasn't invoked yet */
+    uct_worker_progress_unregister_safe(&iface->super.worker->super,
+                                        &self->slow_prog_id);
+
+    pthread_mutex_destroy(&self->ops_mutex);
+    if (!ucs_queue_is_empty(&self->ops)) {
+        ucs_warn("destroying endpoint %p with not completed operations", self);
+    }
+
+    /* mark this ep as destroyed so that arriving events on it won't try to
+     * use it */
+    if (self->sock_id_ctx != NULL) {
+        sock_id_ctx     = self->sock_id_ctx; // FIXME!
+        sock_id_ctx->ep = NULL;
+        ucs_debug("ep destroy: sock_id %d", sock_id_ctx->sock_id);
+    }
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+UCS_CLASS_DEFINE(uct_tcpcm_ep_t, uct_base_ep_t)
+UCS_CLASS_DEFINE_NEW_FUNC(uct_tcpcm_ep_t, uct_ep_t, const uct_ep_params_t *);
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_tcpcm_ep_t, uct_ep_t);
+
+static unsigned uct_tcpcm_client_err_handle_progress(void *arg)
+{
+    uct_tcpcm_ep_t *tcpcm_ep = arg;
+    uct_tcpcm_iface_t *iface = ucs_derived_of(tcpcm_ep->super.super.iface,
+                                               uct_tcpcm_iface_t);
+
+    ucs_trace_func("err_handle ep=%p", tcpcm_ep);
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    tcpcm_ep->slow_prog_id = UCS_CALLBACKQ_ID_NULL;
+    uct_set_ep_failed(&UCS_CLASS_NAME(uct_tcpcm_ep_t), &tcpcm_ep->super.super,
+                      tcpcm_ep->super.super.iface, tcpcm_ep->status);
+
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return 0;
+}
+
+void uct_tcpcm_ep_set_failed(uct_iface_t *iface, uct_ep_h ep, ucs_status_t status)
+{
+    uct_tcpcm_iface_t *tcpcm_iface = ucs_derived_of(iface, uct_tcpcm_iface_t);
+    uct_tcpcm_ep_t *tcpcm_ep       = ucs_derived_of(ep, uct_tcpcm_ep_t);
+
+    if (tcpcm_iface->super.err_handler_flags & UCT_CB_FLAG_ASYNC) {
+        uct_set_ep_failed(&UCS_CLASS_NAME(uct_tcpcm_ep_t), &tcpcm_ep->super.super,
+                          &tcpcm_iface->super.super, status);
+    } else {
+        /* invoke the error handling flow from the main thread */
+        tcpcm_ep->status = status;
+        uct_worker_progress_register_safe(&tcpcm_iface->super.worker->super,
+                                          uct_tcpcm_client_err_handle_progress,
+                                          tcpcm_ep, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                          &tcpcm_ep->slow_prog_id);
+    }
+}
+
+/**
+ * Caller must lock ep->ops_mutex
+ */
+void uct_tcpcm_ep_invoke_completions(uct_tcpcm_ep_t *ep, ucs_status_t status)
+{
+    uct_tcpcm_ep_op_t *op;
+
+    ucs_assert(pthread_mutex_trylock(&ep->ops_mutex) == EBUSY);
+
+    ucs_queue_for_each_extract(op, &ep->ops, queue_elem, 1) {
```

Comment:
We can do:
```
pthread_mutex_lock(&ep->ops_mutex);
ucs_queue_for_each_extract(op, &ep->ops, queue_elem, 1) {
        uct_invoke_completion(op->user_comp, status);
        ucs_free(op);
    }
pthread_mutex_unlock(&ep->ops_mutex);
```
and avoid the requrement for the acquiring lock outside of the function.